pm: device: runtime: API cleanup

gmarull opened the pull request:
While the device runtime API has been available for a long time, it has rarely been used (just a couple of drivers, so probably < 1%). I guess one of the reasons is the lack of documentation or usage guidelines. The aim of this RFC is to clarify API usage (e.g. why should I care about async?), make things a bit simpler and improve the documentation. 

Commit messages provide more details and context on the changes.

Other points to discuss:

- Should asynchronous be optional to save resources (ie force it to be sync depending on a Kconfig setting?)

gmarull committed:
tests: pm: device_runtime: delete test

Delete current device_runtime test. A refactored test will be added in
later commits. It is done at this point to prevent a broken range of
commits due to API changes/removal.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: runtime: remove pm_device_get_async

As of today there is no clear usage of asynchronous gets, since in
general, a resume operation should be synchronous (we are about to use
the device immediately after resuming it). Removing this API simplifies
the runtime implementation in a significant way (refer to future
commits).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: runtime: remove pm_device_wait

After the removal of pm_device_get_async, the pm_device_wait API has
become redundant. Its usage after pm_device_put_async should not be
considered a valid usecase, since after that call what will happen is a
pm_device_get (which is blocking).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: runtime: refactor API

This patch refactors the runtime API to make it more clear and simple.

Relevant changes:

- API uses the action callback in a direct manner, avoiding unnecessary
  overhead.
- API documentation has been improved to include detailed return error
  codes.
- pm_runtime_disable() is now synchronous (to simplify possible error
  paths) and returns error in case it fails. It is also safe to disable
  in pre-kernel now.
- pm_runtime_put(_async)() will return -EALREADY if called with usage
  count at zero (result of an unbalanced get/put call sequence)
- A transitional state has been added back. This makes code more
  readable, and avoids using atomics (not required).

TODO:

- Solve in a better manner the asynchronous suspend error path (now
  "solved" using asserts).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: pm: device_runtime: add test for the API

Add a test for the device runtime API that makes sure most common usage
scenarios behave as expected:

- get + put
- get + asynchronous put until suspended
- get + asynchronous put + get (while suspend still ongoing)

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

yperess reviewed (commented):
Looking at the device_runtime.c implementation I'm wondering if calling get/put should also require passing in a token that would make them effectively re-entrant. Basically, we wouldn't want one bad driver to call put more than get and put the device to sleep. If the caller had to pass a token we could make it re-entrant.

teburd reviewed (commented):
Looks good, waiting on the CI to approve

gmarull committed:
pm: device: runtime: use pm_device_runtime* namespace

Move all PM device runtime API calls from pm_device* to the
pm_device_runtime* namespace.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -1,251 +1,243 @@
+		(void)k_mutex_lock(&pm->lock, K_FOREVER);
+	} else {
+		async = false;
+	}
```
adjusted

gmarull commented on include/pm/device_runtime.h:
```diff
@@ -8,107 +9,104 @@
+ * @retval -errno Other negative errno, result of resuming the device.
  */
-void pm_device_disable(const struct device *dev);
+int pm_device_disable(const struct device *dev);
```
latest commit moves the API to the `pm_device_runtime*` _namespace_

gmarull committed:
tracing: update pm device runtime

Update the tracing definitions used by the device runtime API.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

JordanYates commented:
No complaints here, I use runtime power management extensively out-of-tree and have never found a use for `pm_device_get_async`.

JordanYates reviewed (commented):
No complaints here, I use runtime power management extensively out-of-tree and have never found a use for pm_device_get_async.

erwango commented on include/pm/device_runtime.h:
```diff
@@ -8,107 +9,104 @@
- * will bring up or resume the device if it is in suspended state
- * based on the device usage count. This call is blocked until the
- * device PM state is changed to resume.
+ * This function will increase device usage count. If the device is suspended,
+ * it will be resumed. In case of failure, usage count and device state will be
```
I would reorder the functions effects here:
- First: "If the device is suspended, it will be resumed." > This is the primary effect expected by the caller
- Then: "This function will increase device usage count." > This is mostly interesting to the pm system

gmarull replied:
Adjusted documentation with a better description, let me know if it's better now

erwango commented on include/pm/device_runtime.h:
```diff
@@ -8,107 +9,104 @@
- * Called by a device driver to mark the device as being released.
- * This API asynchronously put the device to suspend state if
- * it not already in suspended state.
+ * This function will decrease usage count. If usage count reaches zero, the
+ * device will be suspended. In case of failure, usage count and device state
```
Ditto: "Suspend the device, based on usage count. Device usage count is decreased. Suspension happens if counter reaches zero"

gmarull committed:
tracing: fix PM system tracing

System PM tracing was broken for SEGGER SystemView, and was missing
proper documentation.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
note: also fixed issues with system PM tracing as CI was complaining

erwango commented on include/pm/device_runtime.h:
```diff
@@ -8,141 +9,118 @@
- * device PM state is changed to resume.
+ * This function will schedule the device suspension if the device is no longer
+ * required (usage count equal to 0). In all other cases, usage count will be
+ * decremented (up to 0).
```
ditto

erwango commented on include/pm/device_runtime.h:
```diff
@@ -8,141 +9,118 @@
+ * This function will suspend the device if the device is no longer required
+ * (usage count equal to 0). In case of suspend failure, usage count and device
+ * state will be left unchanged. In all other cases, usage count will be
+ * decremented (up to 0).
```
"down to 0" ;-)
Otherwise, this is fine to me

gmarull replied:
done

erwango reviewed (approved):
Approved (assuming the CI will eventually pass)
